### DISCUSSION TITLE
x-trap.inert DOM Crawl Issue

### DISCUSSION THREAD

**Discussion by vakisp** [alpinejs/alpine#2506] [7. Bugs (archived)]
Hi All,

I think I may have found an issue with the .inert modifier for the x-trap feature. I am no expert on how the internals of Apline work or for that matter Javascript, so please accept my apologies if this is my mistake.

I noticed when using x-trap.inert on a modal, in some situations it took a very long time for it to show up, after a lot of investigation and debugging it seems that when looking through the DOM to add the aria-hidden attribute it seems to find a lot of nodes, in some cases around 100,000 for a modest page when there are less that 1,000 at most.

I am using this in conjunction with Livewire so thought that may be the issue but I suspect not.

You don't really notice the issue unless you are using chrome with the javascript console open, Safari is aways instant and Chrome is quick unless you have the console open.

Looking at the code (index.js) I find this:

function crawlSiblingsUp(el, callback) {
  if (el.isSameNode(document.body) || !el.parentNode)
    return;
  Array.from(el.parentNode.children).forEach((sibling) => {
    if (!sibling.isSameNode(el))
      callback(sibling);
    **crawlSiblingsUp(el.parentNode, callback);**
  });
}

If I am reading this right crawlSiblingsUp is called recursively to walk up the DOM, but the line **crawlSiblingsUp(el.parentNode, callback)** is called for every sibling, but this means it will just do the same thing every time as el.parentNode will be the same for every iteration of the Array, so if there are 10 siblings it will call crawlSiblingsUp with the same el.parentNode 10 times which causes a massive set nodes to be added to the undo's array?

My initial though is to move crawlSiblingsUp(el.parentNode, callback) outside of the the Array iterator loop, but I am not sure I am confident enough in my coding skills to be sure that is right.

Alternatively I could have got this all wrong and it is either something I am doing wrong, my DOM is somehow broken or it doesn't play nice with Livewire, if so please accept my apologies.

Any help, comments or confirmation this is a bug will be much appreciated.

Thanks,

Vakis

**Comment by danddanddand** [comment]
I don't know the implementation details but we found inert modifier had performance issues in our application and we couldn't use it

**Reply by vakisp** [reply]
I modified the code as suggested above in my local dev environment and it performs much faster and appears to be putting the aria-hidden attributes in the right places. I will see what the general consensus is and if people here think it's a bug I will raise this  this as an issue (I assume that is the correct protocol, I am a bit new to this).

**Reply by danddanddand** [reply]
If you want to submit a PR, I'll test it in my application and review for you

**Reply by vakisp** [reply]
OK happy to do that but I will need to figure out how to do a PR, it might take me a few days as I've not done one before but in the mean time the change is in the original posting of it helps you.

**Comment by philwolstenholme** [comment]
Hi @vak, what version of Alpine are you using?

Since [this PR](https://github.com/alpinejs/alpine/pull/2072) and Alpine [3.4.0](https://github.com/alpinejs/alpine/releases/tag/v3.4.0) we've used https://github.com/focus-trap/focus-trap rather than the `crawlSiblingsUp` code in your issue to manage focus trapping. I think if you update Alpine you might avoid the performance issues that you and @danddanddand are having.

**Reply by SimoTod** [reply]
@philwolstenholme i think the inert modifier has been added back on top of focus trap

**Reply by philwolstenholme** [reply]
Ah yes, you're right! Ignore me for now :)

**Comment by program-mia** [comment]
Hi!
I know it's an old thread but I thought I jump in too. I just tried to add ``x-trap.inert="open"`` to general modal in our app and I have the same issue when opening a modal can have me waiting for up to even a minute for some pages. I can confirm it adds the ``aria-hidden="true"`` everywhere and I am just not sure if there is some way of speeding this process?
I am using alpine 3.9.3 (cdn) together with focus 3.x.x (cdn, latest)
